README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# tr808r (pronounced ‘traitor’) <img src="man/figures/logo-tr808r.png" align="right" width="230"/>

<!-- badges: start -->

![](https://img.shields.io/badge/cool-useless-green.svg)
<!-- badges: end -->

`{tr808r}` is a tiny sound-a-like for [Roland’s TR-808 drum
machine](https://en.wikipedia.org/wiki/Roland_TR-808).

This is a tech demo for how
[`eventloop}`](https://github.com/coolbutuseless/eventloop) allows you
to write complex interactive applications using just R.

## What’s in the box

-   `tr808()` Start an interactive TR-808 session.

## Limitations/Future possibilities

-   No current support for pre-scale selection or accents.
-   No support for any level/tone decay tuning of any instrument.
-   Base instruments only. No congas, claves or maracas.

## Installation

Note that `{tr808r}` relies on the `{eventloop}` package for handling
the interactivity on macOS and unix-like platforms.

Windows OS is not supported by `{eventloop}` as R on Windows does not
(currently) support the `onIdle` callback.

You can install from [GitHub](https://github.com/coolbutuseless/tr808r)
with:

``` r
# install.package('remotes')
remotes::install_github('coolbutuseless/eventloop')  #  >= v0.1.1
remotes::install_github('coolbutuseless/tr808r')
```

## Compose your own

-   Press SPACE to play/pause.
-   Click boxes to toggle playback of that instrument at that time.
-   There are 8 editable patterns.
-   Use keys ‘1’ to ‘8’ to toggle that pattern as active. This will let
    you compose longer rhythms by chaining together multiple patterns.
-   Use SHIFT ‘1’ - ‘8’ to immediately jump to a particular pattern.
-   UP and DOWN arrows will change BPM (Beats Per Minute).
-   Pressing ‘s’ will save the machine state in the current directory in
    a time-stamped file.
    -   This file can be reloaded with `tr808(state = ...)`
-   Pressing ‘t’ will output a text representatin of the current pattern
    which you can tweet!
-   You can play back a tweeted pattern by copying the text from twitter
    and passing it in to `tr808(state = ...)`

``` r
tr808()
```

#### <a href="https://coolbutuseless.github.io/package/tr808r/images/compose.mp4">Watch video with sound</a>

<img src="man/figures/schematic.png" />

## Play back an existing song pattern

Open the drum machine with an existing song.

``` r
tr808(state = demo_songs[[1]])
```

#### <a href="https://coolbutuseless.github.io/package/tr808r/images/demo1.mp4">Watch video with sound</a>

## Tweeting patterns and playing back tweeted patterns

This is an easy, fun(?) way to share patterns on twitter!

#### Create your own tweetable

Press `t` (for twitter) within the application to print a text
representation to your R console.

If you have [`{clipr}`](https://cran.r-project.org/package=clipr)
installed then the text will also be copied to your clipboard for easy
pasting into twitter!

    #RStats #tr808r
    BD xx____xx__x_____
    SD ____x_______x___
    LT ________________
    MT xx______________
    HT ______xx__x__x__
    RS ________________
    CP xx_xx_xx_______x
    CB xx__x_xx_x_x_xx_
    CY ________________
    OH ________________
    CH x_xxx_xxx_xxx_xx

#### Playback a tweet

If you’d like to play a pattern that someone has tweeted, copy the text
and pass it to `tr808()` as the initial `state`

``` r
tr808(state = '
#RStats #tr808r
BD xx____xx__x_____
SD ____x_______x___
LT ________________
MT xx______________
HT ______xx__x__x__
RS ________________
CP xx_xx_xx_______x
CB xx__x_xx_x_x_xx_
CY ________________
OH ________________
CH x_xxx_xxx_xxx_xx')
```

## Related Software

-   [Roland’s online version of the TR-808](https://roland50.studio/)
-   [Another web TR-808](https://io808.com/)
-   [Some classic drum patterns for the
    TR-808](http://808.pixll.de/index.php%20patterns)
-   [TB-303 Bass
    synthesizer](https://en.wikipedia.org/wiki/Roland_TB-303)

## Acknowledgements

-   R Core for developing and maintaining the language.
-   CRAN maintainers, for patiently shepherding packages onto CRAN and
    maintaining the repository


NEWS.md
# tr808r 0.1.1  2022-05-14

* Fixed pattern length bug when loading pattern from a tweet


# tr808r 0.1.0  2022-05-13

* Initial release


R/aaa.R

env <- new.env()


RED    <- rgb(222, 105,  78, maxColorValue = 255)
ORANGE <- rgb(238, 170, 102, maxColorValue = 255)
YELLOW <- rgb(243, 212, 102, maxColorValue = 255)
WHITE  <- rgb(241, 243, 245, maxColorValue = 255)
BLACK  <- rgb( 30,  30,  30, maxColorValue = 255)

ROLAND_RED <- rgb(230, 102, 48, maxColorValue = 255)

LABEL_YELLOW <- rgb(243, 234, 202, maxColorValue = 255)

# colorspace::darken(c(RED, ORANGE, YELLOW, WHITE), 0.9) -> col
DRED    <- '#2a0700'
DORANGE <- '#241300'
DYELLOW <- '#1f1800'
DWHITE  <- '#0e1c26'


pattern_ui_light_cols <- c(rep( RED, 4), rep( ORANGE, 4), rep( YELLOW, 4), rep( WHITE, 4))
pattern_ui_dark_cols  <- c(rep(DRED, 4), rep(DORANGE, 4), rep(DYELLOW, 4), rep(DWHITE, 4))

pattern_ui_light_cols <- rep(pattern_ui_light_cols, 18)
pattern_ui_dark_cols  <- rep(pattern_ui_dark_cols , 18)


R/tweet.R



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a single pattern to a text representation
#'
#' @param pat single pattern
#'
#' @return text representation.  Also copies to clipboard if \code{clipr}
#'         package is installed.
#'
#' @noRd
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
pattern_to_text <- function(pat) {
  # pat <- pattern$pattern[[1]]
  txt <- rep('_', nrow(pat))
  txt[pat$active] <- 'x'
  mat <- matrix(txt, nrow=11, ncol=16, byrow = TRUE)
  mat <- mat[11:1,]

  # Add instrument names as first column
  mat <- cbind(
    c('BD ', 'SD ', 'LT ', 'MT ', 'HT ', 'RS ', 'CP ', 'CB ', 'CY ', 'OH ', 'CH '),
    mat
  )

  res <- paste(apply(mat, 1, paste, collapse = ""), collapse = "\n")

  res <- paste("#RStats #tr808r", res, sep = "\n")

  if (requireNamespace('clipr', quietly = TRUE)) {
    message("Pattern copied to clipboard in text format")
    clipr::write_clip(res)
  }

  res
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Check if a snippet of text appears to be a text repreentation of a pattern
#'
#' @param txt single string
#'
#' @return TRUE if this is parseable as a valid pattern
#'
#' @noRd
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
is_text_pattern <- function(txt) {

  res <- is.character(txt) && length(txt) == 1 && !is.na(txt)
  if (!res) {
    return(FALSE)
  }

  matches <- gregexpr("[x_]{16}", txt)

  return(length(matches[[1]]) == 11)
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a text representation into a \code{tr808_state} object with a single pattern
#'
#' The returned object can be loaded into the drum machine using
#' \code{tr808(state = ...)}
#'
#' @param txt Single string must contain 11 strings of length 16 characters consisting of
#'        only \code{_} and \code{x}.  All other characters will be ignored.
#' @param bpm bpm. default: 96
#'
#' @return \code{tr808_state} object
#'
#' @export
#'
#'
#' @examples
#' text_to_tr808_state(
#' '#RStats #tr808r
#'  BD xx____xx__x_____
#'  SD ____x_______x___
#'  LT ________________
#'  MT xx______________
#'  HT ______xx__x__x__
#'  RS ________________
#'  CP xx_xx_xx_______x
#'  CB xx__x_xx_x_x_xx_
#'  CY ________________
#'  OH ________________
#'  CH x_xxx_xxx_xxx_xx'
#')
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
text_to_tr808_state <- function(txt, bpm = 96) {

  stopifnot(is.character(txt), length(txt) == 1, !is.na(txt))

  matches <- gregexpr("[x_]{16}", txt)

  if (length(matches[[1]]) != 11) {
    stop("Expected 11 sequences of [x|_]")
  }

  strings <- regmatches(txt, matches)[[1]]

  mat <- do.call(rbind, rev(strsplit(strings, "")))

  res <- as.vector(t(mat))

  blank_pattern <- expand.grid(x=1:16, y=1:11, active = FALSE)

  first_pattern <- blank_pattern
  first_pattern$active <- res == 'x'

  pattern <- rep(list(blank_pattern), NPAT)
  pattern[[1]] <- first_pattern

  pattern_set <- logical(NPAT)
  pattern_set[[1]] <- TRUE

  list(
    pattern_set = pattern_set,
    pattern     = pattern,
    bpm         = bpm
  )

}


if (FALSE) {
  text <- '#RStats {tr808r}
BD xx____xx__x_____
SD ____x_______x___
LT ________________
MT xx______________
HT ______xx__x__x__
RS ________________
CP xx_xx_xx_______x
CB xx__x_xx_x_x_xx_
CY ________________
OH ________________
CH x_xxx_xxx_xxx_xx'
}


data-raw/prep-samples.R
## code to prepare `prep-samples` dataset goes here

files <- list.files("data-raw/samples/", full.names = TRUE)
sample_names <- basename(tools::file_path_sans_ext(files))
names(files) <- sample_names
samples <- lapply(files, audio::load.wave)


samples <- samples[c(
  'Bassdrum-01',
  'Snaredrum',
  'Tom L',
  'Tom M',
  'Tom H',
  'Rimshot',
  'Clap',
  'Cowbell',
  'Crash-01',
  'Hat Open',
  'Hat Closed'
)]

sample_names <- c(
  'Bass Drum',
  'Snare Drum',
  'Low Tom',
  'Mid Tom',
  'Hi Tom',
  'Rimshot',
  'Hand Clap',
  'Cowbell',
  'Cymbal',
  'Open Hat',
  'Closed Hat'
)


samples <- rev(samples)
sample_names <- rev(sample_names)

names(samples) <- sample_names



usethis::use_data(samples, internal = TRUE, overwrite = TRUE)



demo_songs <- list()
demo_songs[[1]] <- readRDS("data-raw/New Order - Confusion.rds")
demo_songs[[2]] <- readRDS("data-raw/Marvin Gaye - Sexual Healing.rds")
demo_songs[[3]] <- readRDS("data-raw/Michael Jackson - Beat It Intro.rds")
usethis::use_data(demo_songs, sample_names, internal = FALSE, overwrite = TRUE)


R/tr808.R

globalVariables(c('x', 'active'))



PATTERN_UI_OFFSET <- 0.2
PATTERN_UI_SCALE  <- 0.8
NPAT              <- 8   # 8 patterns

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Draw the tempo/tracer across the top
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
draw_tracer <- function(i) {
  cols <- rep('grey30', 16)
  cols[i] <- 'grey80'
  grid::grid.rect(
    x = (seq(0, 1, length.out = 17)[-17] + 1/32) * PATTERN_UI_SCALE + PATTERN_UI_OFFSET,
    y = 0.975,
    width = 0.045, height = 0.05,
    just = c(0.5, 1),
    gp = grid::gpar(fill = cols)
  )
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Draw the pattern editor UI
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
draw_current_pattern <- function() {

  pattern <- env$pattern[[env$current_pattern_idx]]

  fill <- pattern_ui_dark_cols
  fill[pattern$active] <- pattern_ui_light_cols[pattern$active]

  grid.rect(
    width  = 0.04,
    height = 0.065,
    x = (pattern$x / 16 - 1/32) * PATTERN_UI_SCALE + PATTERN_UI_OFFSET,  # 16 beats
    y = pattern$y / 12 - 1/24,  # 11 instruments
    gp = gpar(fill = fill)
  )

}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Update the state of the pattern based upon mouse click position
# Thic function is called when mouse is clicked, and if it corresponds
# to an element in the pattern deck, then set this value to active
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
update_current_pattern <- function(mouse_x, mouse_y) {

  # Instrument pattern
  pattern <- env$pattern[[env$current_pattern_idx]]

  x <- ceiling((mouse_x - PATTERN_UI_OFFSET) / PATTERN_UI_SCALE * 16)
  y <- ceiling(mouse_y * 11 / (1 - 0.065))

  idx <- pattern$x == x & pattern$y == y

  pattern[idx, 'active'] <- !pattern[idx, 'active']

  env$pattern[[env$current_pattern_idx]] <- pattern
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Draw Pattern Select UI
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
draw_pattern_selection <- function() {

  fills <- c('grey20', 'dodgerblue3')[env$pattern_set + 1L]

  cols <- rep(BLACK, NPAT)
  cols[env$current_pattern_idx] <- 'WHITE'

  y <- c(
    0.95, 0.95, 0.95, 0.95,
    0.92, 0.92, 0.92, 0.92
  ) + 0.014

  grid::grid.rect(
    x = seq_len(NPAT/2) / 48 + 4/48,
    y = y,
    width = 1/55,
    height = 1/36,
    gp = gpar(col = cols, fill = fills, lwd = 3)
  )

  grid::grid.text(
    label = seq_len(NPAT),
    x = seq_len(NPAT/2) / 48 + 4/48,
    y = y,
    gp = gpar(
      fontfamily = 'sans',
      col        = 'white',
      cex        = 1
    )
  )
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Update pattern select
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
update_pattern_selection <- function(idx) {
  # user has pressed a key in "1" - NPAT
  idx <- as.integer(idx)
  env$pattern_set[idx] <- !env$pattern_set[idx]

  # If the user turned off the last valid pattern,
  # then turn on pattern #1
  if (!any(env$pattern_set)) env$pattern_set[1] <- TRUE
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Render the names of the instruments somehow
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
draw_instrument_names <- function() {
  y <- (1:11) / 12 - 1/24

  for (yi in y) {
    grid::grid.roundrect(
      x = 0.02,
      y = yi,
      width = 0.16,
      height = 0.05,
      just =c(0, 0.5),
      r = unit(0.2, 'snpc'),
      gp = gpar(col = NA, fill = LABEL_YELLOW)
    )
  }

  grid::grid.text(
    tr808r::sample_names,
    x = 0.03, y = y,
    hjust = -0.1,
    gp = gpar(
      fontfamily = 'sans',
      col        = 'black',
      cex        = 1.75
    )
  )
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
reset_current_pattern <- function() {

  pattern <- expand.grid(x=1:16, y=1:11, active = FALSE)

  env$pattern[[env$current_pattern_idx]] <- pattern

  draw_current_pattern()
}

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
reset_all_patterns <- function() {

  pattern <- expand.grid(x=1:16, y=1:11, active = FALSE)

  env$pattern <- rep(list(pattern), NPAT)

  draw_current_pattern()
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Render the BPM info
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
draw_bpm <- function() {

  # Rectangle underneath
  grid::grid.roundrect(
    x = 0.01, y = 0.95,
    r = unit(0.1, 'snpc'),
    width = 0.06,
    height = 0.05,
    just = c(0, 0.5),
    gp = gpar(
      col = NA,
      fill = 'pink'
    )
  )

  # BPM text on top
  grid::grid.text(
    env$bpm,
    x = 0.06,
    y = 0.95,
    hjust = 1,
    gp = gpar(
      fontfamily = 'sans',
      col        = 'black',
      cex        = 1.75
    )
  )
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Initialise function for eventloop
#'
#' @return None
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tr808_init <- function() {

  env$playing             <- TRUE
  env$count               <- 0L
  env$pattern             <- NULL
  env$pattern_set         <- logical(NPAT)
  env$pattern_set[1]      <- TRUE
  env$current_pattern_idx <- 1L

  reset_all_patterns()

  if (!is.null(env$init_state)) {
    env$pattern     <- env$init_state$pattern
    env$pattern_set <- env$init_state$pattern_set
    env$current_pattern_idx <- which(env$pattern_set)[[1]]

    env$init_state <- NULL
  }

  # Set background
  grid.rect(gp = gpar(col = NA, fill = BLACK))

  # Draw components
  draw_pattern_selection()
  draw_current_pattern()
  draw_instrument_names()
  draw_bpm()
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Run the frame rate at a faster multiple of the BPM so that the UI
# is more responsive.
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
FPS_MULT <- 4
SHIFT    <- log2(FPS_MULT)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Main TR808 eventloop callback
#'
#' @param event,mouse_x,mouse_y,event_env,... eventloop args
#'
#' @import grid
#' @import audio
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tr808_func <- function(event, mouse_x, mouse_y, event_env, ...) {

  i <- ( bitwShiftR(env$count, SHIFT) %% 16L) + 1L

  # Only 'play' audio every n'th frame
  play_audio <- (env$count %% FPS_MULT) == 0


  # Update patterns on first note of first beat
  if (play_audio && env$playing && i == 1 && env$count != 0) {
    all_pattern_indices <- which(env$pattern_set)
    pattern_indices <- all_pattern_indices[all_pattern_indices > env$current_pattern_idx]
    if (length(pattern_indices) == 0) {
      # Loop around to first index
      next_pattern_idx <- all_pattern_indices[1]
      if (next_pattern_idx != env$current_pattern_idx) {
        env$current_pattern_idx <- next_pattern_idx
        draw_current_pattern()
        draw_pattern_selection()
      }
    } else {
      # Move to next pattern, and update the pattern
      env$current_pattern_idx <- pattern_indices[1]
      draw_current_pattern()
      draw_pattern_selection()
    }
  }



  if (play_audio && env$playing) {
    draw_tracer(i)

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Find instruments active at this time step and play them
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    now <- subset(env$pattern[[env$current_pattern_idx]], x == i & active)
    for (sample_idx in now$y) {
      sample_name <- tr808r::sample_names[[sample_idx]]
      audio::play(samples[[sample_name]])
    }
  }

  if (env$playing) {
    env$count <- env$count + 1L
  }


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Mouse clicks update the active state of the instrument pattern
  # Keys:
  #   c - clear pattern
  #   SPACE = play/pause
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  shifted_nums <- c('!', '@', '#', '$', '%', '^', '&', '*')
  if (!is.null(event)) {
    if (event$type == 'key_press') {
      if (event$str == 'c') {
        reset_current_pattern()
        draw_pattern_selection()
      } else if (event$str == ' ') {
        env$playing <- !env$playing
      } else if (event$str %in% seq_len(NPAT)) {
        update_pattern_selection(event$str)
        draw_pattern_selection()
      } else if (event$str %in% shifted_nums) {
        idx <- which(shifted_nums == event$str)
        env$pattern_set[idx] <- TRUE
        env$current_pattern_idx <- idx
        draw_pattern_selection()
        draw_current_pattern()
      } else if (event$str %in% c('-', '_', 'Down')) {
        env$bpm <- env$bpm - 1L
        event_env$fps_target <- bpm_to_fps(env$bpm)
        draw_bpm()
      } else if (event$str %in% c('+', '=', 'Up')) {
        env$bpm <- env$bpm + 1L
        event_env$fps_target <- bpm_to_fps(env$bpm)
        draw_bpm()
      } else if (event$str == 's') {
        datestamp <- strftime(Sys.time(), "%Y%m%d-%H:%M:%S")
        filename <- paste0('tr808-', datestamp, ".rds")
        message("Saved to: ", filename)
        save_obj <- list(
          pattern_set = env$pattern_set,
          pattern     = env$pattern,
          bpm         = env$bpm
        )
        saveRDS(save_obj, file = filename)
      } else if (event$str == 't') {
        cat(pattern_to_text(env$pattern[[env$current_pattern_idx]]))
      }
    } else if (event$type == 'mouse_down') {
      update_current_pattern(mouse_x, mouse_y)
      draw_current_pattern()
    }

  }


}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Run the FPS at a multiple of the BPM
# E.g. 96 BPM is only 6.4Hz
#      Having a framerate of 6.4Hz makes the UI feel really unresponsive
# So just multiply by a factor, and divide by this factor later on
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bpm_to_fps <- function(bpm) {
  bpm * 4 / 60 * FPS_MULT
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Is this object a valid pattern of 11instruments * 16 notes
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
is_pattern <- function(pattern) {
  is.data.frame(pattern) &&
    nrow(pattern) == 176 &&
    !anyNA(pattern) &&
    all(pattern$x %in% 1:16) &&
    all(pattern$y %in% 1:11) &&
    all(pattern$active %in% c(TRUE, FALSE))
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Is this a valid list of patterns?
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
is_pattern_list <- function(pattern) {
  is.list(pattern) &&
    length(pattern) == NPAT &&
    !anyNA(pattern) &&
    all(
      vapply(pattern, is_pattern, logical(1))
    )
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Is this a valid tr808 state?
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
is_tr808_state <- function(state) {
  is.list(state) &&
  is.logical(state$pattern_set) &&
    !anyNA(state$pattern_set) &&
    any(state$pattern_set) &&
    is_pattern_list(state$pattern) &&
    is.numeric(state$bpm) &&
    !is.na(state$bpm)
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' TR-808
#'
#' @param width,height,... arguments passed to \code{eventloop::run_loop()}
#' @param bpm beats per minute. default: 96
#' @param state a \code{tr808_state} object or the path of a saved object in
#'        an RDS file.  Note: Pressing 's' within the interactive drum machine
#'        window will save the current state to a time stamped RDS file.
#'
#' @importFrom utils packageVersion
#' @import eventloop
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tr808 <- function(bpm = 96, width = 12, height = 8, state = NULL, ...) {

  if (!interactive()) {
    stop("Running this app only makes sense when R is interactive")
  }

  if (packageVersion('eventloop') < "0.1.1") {
    stop("Need {eventloop} v0.1.1\nremotes::install_github('coolbutuseless/eventloop')")
  }


  if (!is.null(state)) {
    if (is_tr808_state(state)) {
      # do nothing
    } else if (file.exists(state)) {
      state <- readRDS(state)
      if (!is_tr808_state(state)) {
        stop("File specified does not contain a tr808 state object")
      }
    } else if (is_text_pattern(state)) {
      state <- text_to_tr808_state(state)
    } else {
      stop("'state' argument must be a tr808 state object, a file path to a ",
           "saved object in RDS format, or a text string coercible to a",
           " pattern")
    }
    env$init_state <- state

    # Use BPM of the loaded object rather than function argument
    bpm <- state$bpm
  }


  env$bpm <- as.integer(bpm)

  fps <- bpm_to_fps(bpm)
  message("BPM: ", bpm, "  FPS: ", fps)
  message("SPACE to play/pause")
  message("'c' to clear current pattern")
  message("'s' to save tr808 state (load with 'state = ...' argument)")
  message("'t' to copy tweetable text to clipboard/print to console")
  message("'1'-'8' to toggle pattern for playback")
  message("SHIFT + '1'-'8' to jump immediately to this pattern (usually best when PAUSED)")
  message("+/- Up/Down to adjust BPM")


  eventloop::run_loop(
    tr808_func, init_func = tr808_init,
    width = width, height = height,
    fps_target = fps, ..., double_buffer = FALSE
  )
}



if (FALSE) {
  library(eventloop)
  run_loop(tr808_func, init_func = tr808_init, bpm = 96)
}
